adapters/repos/db/vector/flat/index.go

```diff
@@ -171,11 +223,17 @@ func (index *flat) Add(id uint64, vector []float32) error {
 	slice := make([]byte, len(vector)*4)
 	index.storeVector(id, byteSliceFromFloat32Slice(vector, slice))
 
-	if index.compression == flatent.CompressionBQ {
+	if index.compression == compressionBQ {
 		vectorBQ, err := index.bq.Encode(vector)
 		if err != nil {
 			return err
 		}
+		if index.bqCache != nil {
+			index.Lock()
+			index.bqCache.Grow(id)
+			index.bqCache.Preload(id, vectorBQ)
+			index.Unlock()
```

Comment:
is locking necessary? cache has couple of internal locks already?

Reply:
BTW: I think cache itself needs some improvement.
Method [Grow](https://github.com/weaviate/weaviate/blob/flat_config/adapters/repos/db/vector/cache/sharded_lock_cache.go#L212-L226) should verify again whether grow is necessary after obtaining all locks. Otherwise it could shrink cache slice that was extended by 2nd call of the method while 1st one was waiting for the locks.

[check](https://github.com/weaviate/weaviate/blob/flat_config/adapters/repos/db/vector/cache/sharded_lock_cache.go#L212-L226) should be copied after `defer s.releaseAllLocks()`

Reply:
in regards to BTW: it was fixed in separate PR https://github.com/weaviate/weaviate/pull/3808

Reply:
You might be right. I had to lock to avoid data corruption at some point but I think I did not fix the problem at its root. I am removing the locks now.

Reply:
still not removed :)